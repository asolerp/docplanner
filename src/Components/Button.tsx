import {useTheme} from '@/Theme';

import React from 'react';
import {TouchableOpacity, Text, GestureResponderEvent} from 'react-native';

interface Props {
  title: string;
  disabled: boolean;
  onPress?: (event: GestureResponderEvent) => void;
}

const Button = ({title, disabled, onPress}: Props) => {
  const {Layout, Fonts, Colors, Gutters, MetricsSizes} = useTheme();
  return (
    <TouchableOpacity
      disabled={disabled}
      onPress={onPress}
      style={[
        Layout.rowCenter,
        Gutters.tinyVPadding,
        Gutters.regularHPadding,
        Gutters.tinyHMargin,
        {
          backgroundColor: Colors.primary,
          borderRadius: MetricsSizes.small,
          opacity: disabled ? 0.3 : 1,
        },
      ]}>
      <Text style={[Fonts.textRegular, {color: Colors.white}]}>{title}</Text>
    </TouchableOpacity>
  );
};

export default Button;
