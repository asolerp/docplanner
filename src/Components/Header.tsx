import {useNavigation} from '@react-navigation/core';
import React from 'react';
import {View, Text, TouchableOpacity} from 'react-native';
import {useTheme} from '@/Theme';

const Header = () => {
  const {Layout, Colors, Fonts, Gutters} = useTheme();
  const navigation = useNavigation();
  return (
    <View
      style={[
        Layout.boxShadow,
        Layout.fullWidth,
        Layout.row,
        {backgroundColor: Colors.white},
        Gutters.regularVPadding,
        Gutters.regularLPadding,
      ]}>
      <TouchableOpacity onPress={() => navigation.pop()}>
        <Text style={[Fonts.textRegular]}>Go back</Text>
      </TouchableOpacity>
    </View>
  );
};

export default Header;
