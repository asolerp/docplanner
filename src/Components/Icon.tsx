import {useTheme} from '@/Theme';
import React from 'react';
import {TouchableOpacity, Text, GestureResponderEvent} from 'react-native';

interface Props {
  symbol: string;
  onPress?: (event: GestureResponderEvent) => void;
}

const Icon = ({symbol, onPress}: Props) => {
  const {Layout, Gutters, Colors} = useTheme();
  return (
    <TouchableOpacity
      onPress={onPress}
      style={[
        Layout.rowCenter,
        Gutters.tinyVPadding,
        Gutters.tinyHPadding,
        {
          backgroundColor: Colors.white,
          borderRadius: 100,
          width: 30,
          height: 30,
        },
      ]}>
      <Text style={{fontSize: 15}}>{symbol}</Text>
    </TouchableOpacity>
  );
};

export default Icon;
