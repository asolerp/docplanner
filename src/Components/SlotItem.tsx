import {format} from 'date-fns';
import React from 'react';
import {useTheme} from '@/Theme';

import {View, Text, TouchableOpacity} from 'react-native';
import {useNavigation} from '@react-navigation/core';
import {useSelector} from 'react-redux';
import {books} from '@/Store/Calendar/calendarSlice';
import {slotIsTaken} from '@/Utils/slots';

interface Props {
  Start: string;
  End: string;
  Taken?: boolean;
}

const SlotItem = ({slot}: Props) => {
  const {Layout, Gutters, Colors} = useTheme();
  const navigation = useNavigation();
  const booksState = useSelector(books);

  const handleClickSlot = slot => {
    navigation.navigate('Slot', {slot});
  };

  return (
    <View style={[Layout.row, Gutters.smallVPadding]}>
      <View style={[{flex: 1}, Layout.rowCenter]}>
        <Text style={{color: Colors.text}}>
          {format(new Date(slot.Start), 'p')}
        </Text>
      </View>
      {slotIsTaken(slot, booksState) ? (
        <TouchableOpacity
          onPress={() => handleClickSlot(slot)}
          style={[
            Gutters.regularVPadding,
            Gutters.smallHPadding,
            {
              flex: 2,
              backgroundColor: Colors.active,
              borderWidth: 1,
              borderColor: Colors.border,
              borderRadius: 5,
              borderLeftWidth: 3,
              borderLeftColor: Colors.booked,
            },
          ]}>
          <Text style={{color: Colors.link}}>Slot booked</Text>
        </TouchableOpacity>
      ) : (
        <TouchableOpacity
          onPress={() => handleClickSlot(slot)}
          style={[
            Gutters.regularVPadding,
            Gutters.smallHPadding,
            {
              flex: 2,
              backgroundColor: Colors.white,
              borderWidth: 1,
              borderColor: Colors.border,
              borderRadius: 5,
            },
          ]}>
          <Text style={{color: Colors.link}}>+ Book Slot</Text>
        </TouchableOpacity>
      )}
    </View>
  );
};

export default React.memo(SlotItem);
