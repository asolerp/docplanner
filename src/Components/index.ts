export {default as WeekNavigator} from './WeekNavigator';
export {default as SlotsViewer} from './SlotsViewer';
export {default as SlotItem} from './SlotItem';
export {default as Header} from './Header';
export {default as Button} from './Button';
export {default as Icon} from './Icon';
