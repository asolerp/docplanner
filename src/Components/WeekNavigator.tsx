import React from 'react';
import {View, Text} from 'react-native';
import {selectWeek} from '@/Store/Calendar/calendarSlice';
import {shallowEqual, useSelector} from 'react-redux';
import {useCalendar} from '@/Components/hooks';
import {format} from 'date-fns';
import {useTheme} from '@/Theme';
import {Icon} from '@/Components';

const WeekNavigator = () => {
  const {Layout, Colors, Gutters, Fonts} = useTheme();
  const weekSelector = useSelector(selectWeek, shallowEqual);
  const {nextWeek, prevWeek} = useCalendar();

  return (
    <View
      style={[
        Layout.fullWidth,
        Layout.rowCenter,
        Gutters.smallVPadding,
        Layout.justifyContentSpaceAround,
        {backgroundColor: Colors.primary},
      ]}>
      <Icon symbol="<" onPress={() => prevWeek()} />
      <View style={[Layout.row]}>
        <Text style={[Fonts.textWhite]}>
          {format(weekSelector.firstDay, 'd')}
        </Text>
        <Text style={[Fonts.textWhite]}>-</Text>
        <Text style={[Fonts.textWhite]}>
          {format(weekSelector.lastDay, 'd MMMM')}
        </Text>
      </View>
      <Icon symbol=">" onPress={() => nextWeek()} />
    </View>
  );
};

export default React.memo(WeekNavigator);
