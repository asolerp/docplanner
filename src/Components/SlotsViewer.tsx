import React, {useCallback, useEffect} from 'react';
import {
  View,
  Text,
  StyleSheet,
  Dimensions,
  FlatList,
  SafeAreaView,
} from 'react-native';
import {format} from 'date-fns';

import {groupedSlots, parseDate} from '@/Utils/slots.js';
import {useTheme} from '@/Theme';

import {
  daysSlots,
  getDaySlots,
  firstDayWeek,
} from '@/Store/Calendar/calendarSlice';

import {useDispatch, useSelector} from 'react-redux';
import SlotItem from './SlotItem';

const styles = StyleSheet.create({
  container: {
    width: Dimensions.get('window').width,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

const SlotsViewer = () => {
  const dispatch = useDispatch();
  const {Layout, Gutters, Colors} = useTheme();
  const firstDayWeekSelector = useSelector(firstDayWeek);
  const slots = useSelector(daysSlots);

  const getSlotsHandler = useCallback(() => {
    dispatch(getDaySlots(format(firstDayWeekSelector, 'yyyyMMdd')));
  }, [dispatch, firstDayWeekSelector]);

  useEffect(() => {
    getSlotsHandler();
  }, [firstDayWeekSelector, getSlotsHandler]);

  const renderSlot = ({item}) => {
    return <SlotItem slot={item} />;
  };

  const renderDay = ({item: [key, value]}) => {
    return (
      <View style={{flex: 1, width: '100%'}} key={key}>
        <View
          style={[
            Layout.fill,
            Gutters.smallHPadding,
            Gutters.smallVPadding,
            {
              backgroundColor: Colors.white,
            },
            Layout.boxShadow,
          ]}>
          <Text
            style={{
              flex: 1,
              fontSize: 20,
              fontWeight: 'bold',
            }}>
            {format(parseDate(key), 'EEEE, d MMMM')}
          </Text>
        </View>
        <FlatList
          data={value}
          keyExtractor={(item, i) => i}
          renderItem={renderSlot}
        />
      </View>
    );
  };

  return (
    <SafeAreaView style={{flex: 1, width: '100%'}}>
      {slots && (
        <FlatList
          style={{width: '100%'}}
          data={Object.entries(groupedSlots(slots))}
          keyExtractor={(item, i) => i}
          renderItem={renderDay}
        />
      )}
    </SafeAreaView>
  );
};

export default React.memo(SlotsViewer);
