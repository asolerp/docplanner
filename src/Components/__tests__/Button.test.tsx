import React from 'react';
import {Button} from '@/Components';
import {render, fireEvent} from '@testing-library/react-native';
import renderer from 'react-test-renderer';

const mockPress = jest.fn();

test('renders correctly', () => {
  const tree = renderer
    .create(<Button title="Botón" onPress={mockPress} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('button with custom title emits onPress event', () => {
  const {getByText} = render(<Button title="Botón" onPress={mockPress} />);
  const button = getByText('Botón');

  fireEvent.press(button);

  expect(mockPress).toHaveBeenCalled();
});
