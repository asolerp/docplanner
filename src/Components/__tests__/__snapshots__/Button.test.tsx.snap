// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#40C3A5",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginHorizontal": 5,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 5,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#747474",
          "fontSize": 20,
        },
        Object {
          "color": "#FFFFFF",
        },
      ]
    }
  >
    Bot√≥n
  </Text>
</View>
`;
